### PR TITLE
[cache_filter] Separate upstream from downstream

### PR DESCRIPTION
**Pull request by ravenblackx**
Commit Message: [cache_filter] Separate upstream from downstream
Additional Description: A recurring issue with cache_filter is that because it uses the regular upstream from a client's connection to populate the cache, a client being slow to download, or disconnecting mid-stream, can cause the cache-populating upstream to be delayed by watermarks or aborted. If the cache were to attempt to stream the results while they're still being populated, this can lead to many clients all receiving an unexpected disconnect or an unnecessarily slow response.
This PR makes the cache filter use a separate upstream made with AsyncHttpClient, so even if the client disconnects the upstream can persist and continue to populate the cache entry.

There are two flaws in this version.
1. the downstream write and the cache write are still being performed in parallel so if one is slower than the other the slower one's buffer can become oversized in memory
2. the upstream read has no speed limit so if the cache write is slower than the network read then the cache write buffer can become oversized in memory.

Subsequent PRs are intended to address these:
1. once the cache provides a "streamed" interface from the cache while the entry is being written, we can use that for the "primary" downstream the same as any other downstream requesting the same resource, rather than streaming from the "shared read" as it does now.
2. adding an API for the ability to pause and resume an AsyncHttpClient Stream's "read" interface so if the cache-write is a bottleneck the amount of data buffered in memory is capped.
3. cache filter also needs to monitor downstream watermark callbacks to avoid pushing data faster than the downstream consumes it.

Risk Level: Low, filter is WIP and previously bug-ridden and crash-prone. Integration test passes with no changes after this change.
Testing: Mostly covered by existing tests - one test changed from "expect abort" to "expect cache write completion", as is the intent of this PR.
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: api/envoy/extensions/filters/http/cache/v3/cache.proto [hunk 1]
```diff
@@ -52,6 +52,14 @@ message CacheConfig {
 
   // Config specific to the cache storage implementation. Required unless ``disabled``
   // is true.
+  //
+  // Note that when the cache is enabled, cacheable requests are only sent through
+  // filters in the ``upstream_filters`` chain and *not* through any filters in the
+  // regular filter chain that are further upstream than the cache filter, while
+  // non-cacheable requests still go through the listener filter chain. It is
+  // recommended that only the router filter should be further upstream in the
+  // listener filter chain than the cache filter.
```
Comment (adisuissa) [inline api/envoy/extensions/filters/http/cache/v3/cache.proto:61]: Not sure I fully understand.
Do `upstream_filters` refer to [upstream_http_filters](https://github.com/envoyproxy/envoy/blob/aa195eb01d1211aa9e082f2d742d75395975edc0/api/envoy/extensions/filters/http/router/v3/router.proto#L136)?

If so, it is unclear to me why a request skips the downstream-filters and only goes through the upstream filters.

Just a suggestion to make this a bit more understandable: it would be nice to have a diagram that depicts which filters are invoked and which are not (similar to https://www.envoyproxy.io/docs/envoy/latest/intro/life_of_a_request#http-filter-chain-processing where there are arrows that skip some filters.

**Review by adisuissa — COMMENTED** [review]
Thanks! Left a minor API comment.

File: api/envoy/extensions/filters/http/cache/v3/cache.proto (hunk 1 above)
Comment (ravenblackx) [reply api/envoy/extensions/filters/http/cache/v3/cache.proto]: Yes, `upstream_http_filters`. It's not that it skips the downstream-filters, it's that it skips the filter chain in between the router filter and the cache filter, because that's just what the `AsyncHttpRequest` interface does - it's not an intentional behavior, it's an unfortunate concession to how things work internally.

Moved this into the cache rst document because I don't think we can do images links in generated protodoc, and it's probably the more appropriate place for it anyway, and clarified and linked and added an svg.

Comment (ravenblackx) [reply api/envoy/extensions/filters/http/cache/v3/cache.proto]: (Also updated an outdated statement in that doc.)

File: source/extensions/filters/http/cache/cache_filter.cc [hunk 2]
```diff
@@ -117,115 +185,153 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
...
-    ENVOY_BUG(Http::Utility::getResponseStatus(headers) !=
-                  Envoy::enumToInt(Http::Code::RequestTimeout),
-              "Request timed out while cache lookup was outstanding.");
-    filter_state_ = FilterState::NotServingFromCache;
-    return Http::FilterHeadersStatus::Continue;
+  if (lookup_) {
+    lookup_->onDestroy();
+    lookup_ = nullptr;
   }
+  if (insert_queue_) {
```
Comment (toddmgreer) [inline source/extensions/filters/http/cache/cache_filter.cc:200]: Does insert_queue_ guarantee that all callbacks happen on this thread?

**Review by toddmgreer — CHANGES_REQUESTED** [review]

File: source/extensions/filters/http/cache/cache_filter.h [hunk 3]
```diff
@@ -80,21 +82,97 @@ class CacheFilter : public Http::PassThroughFilter,
...
   Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap& headers,
                                           bool end_stream) override;
-  Http::FilterDataStatus encodeData(Buffer::Instance& buffer, bool end_stream) override;
-  Http::FilterTrailersStatus encodeTrailers(Http::ResponseTrailerMap& trailers) override;
 
   static LookupStatus resolveLookupStatus(absl::optional<CacheEntryStatus> cache_entry_status,
                                           FilterState filter_state);
 
 private:
+  class UpstreamRequest : public Http::AsyncClient::StreamCallbacks, public InsertQueueCallbacks {
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.h:90]: Done.

File: source/extensions/filters/http/cache/cache_filter.h [hunk 4]
```diff
@@ -80,21 +82,97 @@ class CacheFilter : public Http::PassThroughFilter,
...
+  public:
+    void sendHeaders(Http::RequestHeaderMap& request_headers);
+    void disconnectFilter();
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.h:93]: Done.

File: source/extensions/filters/http/cache/cache_filter.h [hunk 5]
```diff
@@ -80,21 +82,97 @@ class CacheFilter : public Http::PassThroughFilter,
...
+
+    // InsertQueueCallbacks
+    void insertQueueOverHighWatermark() override;
+    void insertQueueUnderLowWatermark() override;
+    void insertQueueAborted() override;
+
+    static UpstreamRequest* create(CacheFilter* filter, std::shared_ptr<HttpCache> cache,
+                                   Http::AsyncClient& async_client,
+                                   const Http::AsyncClient::StreamOptions& options);
+    UpstreamRequest(CacheFilter* filter, std::shared_ptr<HttpCache> cache,
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.h:110]: That's annoying with smart pointers, unfortunately. (Yeah you can do a `unique_ptr` wrapped around a `new` instead of a `make_unique`, I think that's more problematic than a public constructor in what is essentially a private context anyway.)

File: source/extensions/filters/http/cache/cache_filter.h [hunk 6]
```diff
@@ -80,21 +82,97 @@ class CacheFilter : public Http::PassThroughFilter,
...
+                    Http::AsyncClient& async_client,
+                    const Http::AsyncClient::StreamOptions& options);
+    ~UpstreamRequest() override;
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.h:113]: Same thing about the `unique_ptr`. I don't want it to be a mistake to use a smart pointer.

File: source/extensions/filters/http/cache/cache_filter.h [hunk 7]
```diff
@@ -80,21 +82,97 @@ class CacheFilter : public Http::PassThroughFilter,
...
+    CacheFilter* filter_ = nullptr;
+    LookupResultPtr lookup_result_;
+    LookupContextPtr lookup_;
+    bool is_head_request_;
+    bool request_allows_inserts_;
+    std::shared_ptr<const CacheFilterConfig> config_;
+    FilterState filter_state_;
+    std::shared_ptr<HttpCache> cache_;
+    Http::AsyncClient::Stream* stream_ = nullptr;
+    std::unique_ptr<UpstreamRequest> self_ownership_;
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.h:146]: I think we have linters that aren't big fans of bare pointers, plus *I* am not a fan of bare pointers (I think the ownership semantics are much clearer with explicitly `self_ownership`), plus in one of the later iterations it's going to become a smart pointer that *isn't* owned by itself so it's more convenient to not have it be a bare pointer temporarily and then have to rejigger it. :)

File: source/extensions/filters/http/cache/cache_filter.h [hunk 8]
```diff
@@ -142,13 +213,15 @@ class CacheFilter : public Http::PassThroughFilter,
   // being cancelled.
   InsertStatus insertStatus() const;
 
-  // insert_queue_ ownership may be passed to the queue itself during
-  // CacheFilter::onDestroy, allowing the insert queue to outlive the filter
-  // while the necessary cache write operations complete.
-  std::unique_ptr<CacheInsertQueue> insert_queue_;
+  // upstream_request_ belongs to the object itself, so that it can be disconnected
+  // from the filter and still complete the cache-write in the event that the
+  // downstream disconnects. The filter and the UpstreamRequest must communicate to
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.h:218]: Done.

File: source/extensions/filters/http/cache/cache_insert_queue.cc [hunk 9]
```diff
@@ -69,22 +69,24 @@ class CacheInsertFragmentTrailers : public CacheInsertFragment {
...
   // While zero isn't technically true for the size of headers, headers are
   // typically excluded from the stream buffer limit.
+  fragment_in_flight_ = true;
   insert_context_->insertHeaders(
       response_headers, metadata,
       [this, end_stream](bool cache_success) { onFragmentComplete(cache_success, end_stream, 0); },
       end_stream);
-  fragment_in_flight_ = true;
+  ASSERT(fragment_in_flight_,
+         "insertHeaders must post the callback to dispatcher, not just call it");
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_insert_queue.cc:89]: The assumption is that *most* cache implementations will need to do something asynchronous, which means they'll be posting. It doesn't *specifically* care, but it would require implementing distinct code paths and more, different guards and things to ensure API correctness if direct callbacks were accepted.

This is the same problem many of the filter functions have, like you can't call `continueDecoding()` in a callback that *might be* called directly during `decodeHeaders()`. A previous cache implementation explicitly posted from within the callback to avoid that case; now, to avoid double-posting in the common case of a cache doing something async, the filter requires the cache to post, which is a performance penalty for the case where the cache is synchronous as in simple_http_cache, but should be an improvement in the common case.

(It's possible that the post requirement can go away because we're not doing continueDecoding any more, but I'd rather not change the API requirement until the rest of this is resolved and I'm sure it won't be important again.)

File: source/extensions/filters/http/cache/cache_insert_queue.cc [hunk 10]
```diff
@@ -135,26 +141,29 @@ void CacheInsertQueue::onFragmentComplete(bool cache_success, bool end_stream, s
...
+      if (callbacks_.has_value()) {
+        callbacks_->insertQueueUnderLowWatermark();
       }
       watermarked_ = false;
     }
     fragments_.clear();
     // Clearing self-ownership might provoke the destructor, so take a copy of the
     // abort callback to avoid reading from 'this' after it may be deleted.
-    auto abort_callback = std::move(abort_callback_);
+    auto callbacks = std::move(callbacks_);
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_insert_queue.cc:161]: Added more comment explaining why that would be a mistake. (It's likely this is one of the things that will be tidier after shared-cache-streaming responses make the ownership semantics less tangled.)

File: source/extensions/filters/http/cache/cache_insert_queue.h [hunk 11]
```diff
@@ -50,8 +54,7 @@ class CacheInsertQueue {
   Event::Dispatcher& dispatcher_;
   const InsertContextPtr insert_context_;
   const size_t low_watermark_bytes_, high_watermark_bytes_;
-  OptRef<Http::StreamEncoderFilterCallbacks> encoder_callbacks_;
-  AbortInsertCallback abort_callback_;
+  OptRef<InsertQueueCallbacks> callbacks_;
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_insert_queue.h:57]: `Anything*` is always simpler than `OptRef<Anything>`, so this comment is essentially "OptRef shouldn't exist".

The value is `OptRef` adds linter assertions that you don't use a potential `nullptr`.

File: source/extensions/filters/http/cache/cache_filter.cc [hunk 12]
```diff
@@ -117,115 +185,153 @@ Http::FilterHeadersStatus CacheFilter::decodeHeaders(Http::RequestHeaderMap& hea
...
+  filter_ = nullptr;
+  if (insert_queue_ == nullptr) {
+    abort();
   }
+}
 
+void CacheFilter::UpstreamRequest::onHeaders(Http::ResponseHeaderMapPtr&& headers,
+                                             bool end_stream) {
+  if (filter_state_ == FilterState::ValidatingCachedResponse && isResponseNotModified(*headers)) {
+    return processSuccessfulValidation(std::move(headers));
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.cc]: I asked about this too, and was told the AsyncHttpClient guarantees that no other callbacks are called after `reset()`. Though hm, I'm not certain if that means `onReset` doesn't get called either, we may need to call it ourselves. Maybe I should validate this behavior in an integration test.

File: source/extensions/filters/http/cache/cache_filter.cc [hunk 13]
```diff
@@ -56,19 +57,86 @@ void CacheFilter::onDestroy() {
...
+  insert_queue_ = nullptr;
+  ENVOY_LOG(debug, "cache aborted insert operation");
+  setInsertStatus(InsertStatus::InsertAbortedByCache);
+  if (filter_ == nullptr) {
+    abort();
   }
 }
 
+void CacheFilter::UpstreamRequest::sendHeaders(Http::RequestHeaderMap& request_headers) {
+  stream_->sendHeaders(request_headers, true);
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.cc]: Done.

File: source/extensions/filters/http/cache/cache_filter.cc [hunk 14]
```diff
@@ -56,19 +57,86 @@ void CacheFilter::onDestroy() {
...
+}
+
+CacheFilter::UpstreamRequest::UpstreamRequest(CacheFilter* filter, std::shared_ptr<HttpCache> cache,
+                                              Http::AsyncClient& async_client,
+                                              const Http::AsyncClient::StreamOptions& options)
+    : filter_(filter), lookup_result_(std::move(filter_->lookup_result_)),
+      lookup_(std::move(filter->lookup_)), is_head_request_(filter->is_head_request_),
+      request_allows_inserts_(filter->request_allows_inserts_), config_(filter->config_),
+      filter_state_(filter->filter_state_), cache_(std::move(cache)),
+      stream_(async_client.start(*this, options)) {
```
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.cc]: Yes, the async clients are inherently thread-local.

File: source/extensions/filters/http/cache/cache_filter.cc (hunk 2 above)
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.cc]: Kind of, that's why everything has to be dispatcher-posted. :)

File: source/extensions/filters/http/cache/cache_filter.cc (hunk 12 above)
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.cc]: Turns out there is already an integration test that covers this path, and I *think* it would complain in one of the *san builds if a memory leak was provoked, which would be the effect if `onReset` didn't get called, so we have a guarantee there. (I also, to be on the safe side, tested in the integration test with some logging to make sure it was behaving as expected.)

**Comment by jmarantz** [comment]
Todd once you are good with this I will review.

**Comment by jmarantz** [comment]
It does look like Raven responded to all of @toddmgreer 's comments so once he's good with that I'll go through it.

**Comment by toddmgreer** [comment]
I've been slammed w/ a critical deadline, but should be able to respond
again either today or tomorrow

File: source/extensions/filters/http/cache/cache_filter.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (toddmgreer) [reply source/extensions/filters/http/cache/cache_filter.h]: I'm very sorry for the painfully slow replies--I should be done digging out from my problems elsewhere, and should be able to reply in a timely manner for the foreseeable future.

I didn't intend to suggest replacing anything with a bare pointer (though bare pointers are the right tool for many situations). My intent was to suggest
1. remove `std::unique_ptr<UpstreamRequest> self_ownership_;`
2. Replace the body of UpstreamRequest::create with `return new UpstreamRequest(...);`
3. Replace the body of UpstreamRequest::onReset, with `delete this`
4. In UpstreamRequest::onComplete, replace `self_ownership_ = nullptr` with `delete this`

I have several reasons for suggesting this:
1. The standard meaning of a `unique_ptr<Foo> foo_` member is "when this class is destroyed, that will in turn delete the pointer held by foo_." self_ownership_ inverts that meaning, which is counter to normal expectations.
2. `delete this` is a well-known pattern. In particular, it's reasonable to expect C++ programmers to know that it would be bad to touch any member variables after `delete this`. Even new C++ programmers should at least see `delete this` and think "something weird is happening." In comparison, `self_ownership_ = nullptr` just looks like a manipulation of a member, and doesn't raise any obvious danger flags. 
3. Getting rid of `self_ownership_` means one fewer member variable to understand, and simplifies the ownership structure.

If a future PR will introduce a real need for a smart pointer, I suspect it will be as convenient to add it then as to replace self_ownership_ with whatever it's to become.

File: source/extensions/filters/http/cache/cache_insert_queue.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (toddmgreer) [reply source/extensions/filters/http/cache/cache_insert_queue.h:57]: IIUC, you're saying that e.g.
`callbacks_->insertQueueUnderLowWatermark()` would hit a lint warning if it weren't guarded by
`if (callbacks_.has_value())`. Is that right? Thanks--I was wondering what OptRef's value was.

Regardless, I shouldn't've raised the question in this PR, because the code already had OptRef here--sorry.

**Review by toddmgreer — CHANGES_REQUESTED** [review]
The comment I just left regarding self_ownership_ is my only concern regarding this PR.

**Comment by tyxia** [comment]
/wait

Per latest comment https://github.com/envoyproxy/envoy/pull/36309#pullrequestreview-2368410959

File: source/extensions/filters/http/cache/cache_filter.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.h]: Done. Disagree but don't care that much.

**Comment by ravenblackx** [comment]
> The comment I just left regarding self_ownership_ is my only concern regarding this PR.

Addressed. Time for @jmarantz !

**Comment by jmarantz** [comment]
ack: will plan to look tomorrow morning.

**Comment by ravenblackx** [comment]
Pre-emptive request, I realize there's a lot of undocumented functions in this PR - *so much* of this is going to be removed or drastically altered in the next round, so my request is, can I leave doc-comments until later when the API is all settled down, to avoid the wasted effort it would be to doc-comment now?

**Review by toddmgreer — APPROVED** [review]
Thank you!

File: source/extensions/filters/http/cache/cache_filter.cc [hunk 15]
```diff
@@ -56,19 +54,41 @@ void CacheFilter::onDestroy() {
...
   }
 }
 
+void CacheFilter::sendUpstreamRequest(Http::RequestHeaderMap& request_headers) {
+  Router::RouteConstSharedPtr route = decoder_callbacks_->route();
+  const Router::RouteEntry* route_entry = (route == nullptr) ? nullptr : route->routeEntry();
+  if (route_entry == nullptr) {
+    return sendNoRouteResponse();
+  }
+  const auto thread_local_cluster =
```
Comment (jmarantz) [inline source/extensions/filters/http/cache/cache_filter.cc:69]: prefer not using 'auto' per style guide. In particular as a reader I don't know if this is an object, ref, shared_ptr,  unique_ptr, or raw ptr.

File: source/extensions/filters/http/cache/cache_insert_queue.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/extensions/filters/http/cache/cache_insert_queue.cc:89]: this is a great PR comment. Can you summarize this as a code comment also?

File: source/extensions/filters/http/cache/cache_insert_queue.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/extensions/filters/http/cache/cache_insert_queue.h:57]: @toddmgreer : fyi OptRef also has these benefits
 * a pointer might be the start of an array, but OptRef cannot
 * a pointer might be owned by the object containing it, but OptRef cannot
 * a pointer might be intended to never be null, but OptRef directly conveys programmer intent that has_value==false is a normal expectation.

Thus its narrower use-case conveys user intent more precisely than pointers in those three cases.

I will also note that OptRef was introduced to wrap  verbose uses with `optional<std::reference<T>>` with terser code. The verbose form was somewhat common in Envoy before, as raw pointers were discouraged in the style guide, so OptRef was a cleaner way of expressing that. Later OptRef's impl was changed to a raw pointer, reducing memory overhead compared to `optional<std::ref...>`

**Review by jmarantz — COMMENTED** [review]
lgtm with small nits.

I'm wondering if @tyxia should also look.

File: source/extensions/filters/http/cache/cache_filter.cc (hunk 15 above)
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_filter.cc]: Done.

File: source/extensions/filters/http/cache/cache_insert_queue.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (ravenblackx) [reply source/extensions/filters/http/cache/cache_insert_queue.cc:98]: Done. (Though the summary is almost as long.)

**Review by jmarantz — APPROVED** [review]

**Comment by triplewy** [comment]
@ravenblackx Does this MR skip all the router filter processing (i.e. header mutation) for cache upstream requests?
